ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/pull-queries-against-ctas-materialized.json

```diff
@@ -205,6 +205,451 @@
+      ],
+      "responses": [
+        {"admin": {"@type": "currentStatus"}},
+        {"admin": {"@type": "currentStatus"}},
+        {"query": [
+          {"header":{"schema":"`GRADE` STRING KEY, `NUM_STUDENTS` BIGINT"}},
+          {"row":{"columns":["D",1]}},
+          {"row":{"columns":["A",3]}},
+          {"row":{"columns":["F",1]}}
+
+        ]},
+        {"query": [
+          {"header":{"schema":"`GRADE` STRING KEY, `NUM_STUDENTS` BIGINT"}},
+          {"row":{"columns":["D",1]}},
+          {"row":{"columns":["F",1]}}
+
+        ]}
+      ]
+    },
+    {
+      "name": "persistent query with COMPLEX filter on all columns + GROUPBY + HAVING + projection +++ pull query table scan and single key lookup",
+      "statements": [
+        "CREATE TABLE INPUT (ID INT PRIMARY KEY, GRADE STRING, RANK INT) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE TABLE NEW_INPUT AS SELECT COUNT(ID) AS NUM_STUDENTS, GRADE FROM INPUT WHERE (ID > 12 AND GRADE IN ('A', 'D')) OR RANK < 2 GROUP BY GRADE HAVING COUNT(ID) > 2;",
+        "SELECT * FROM NEW_INPUT;",
+        "SELECT * FROM NEW_INPUT WHERE NUM_STUDENTS=4;",
+        "SELECT GRADE FROM NEW_INPUT WHERE NUM_STUDENTS=4;"
+
+      ],
+      "inputs": [
+        {"topic": "test_topic", "timestamp": 12333, "key": 9, "value": {"GRADE": "F", "RANK": 0}},
+        {"topic": "test_topic", "timestamp": 12346, "key": 11, "value": {"GRADE": "A", "RANK": 1}},
+        {"topic": "test_topic", "timestamp": 12345, "key": 10, "value": {"GRADE": "B", "RANK": 2}},
+        {"topic": "test_topic", "timestamp": 12389, "key": 12, "value": {"GRADE": "C", "RANK": 3}},
+        {"topic": "test_topic", "timestamp": 12397, "key": 14, "value": {"GRADE": "A", "RANK": 4}},
+        {"topic": "test_topic", "timestamp": 12478, "key": 15, "value": {"GRADE": "B", "RANK": 5}},
+        {"topic": "test_topic", "timestamp": 12378, "key": 16, "value": {"GRADE": "A", "RANK": 6}},
+        {"topic": "test_topic", "timestamp": 12378, "key": 18, "value": {"GRADE": "D", "RANK": 7}},
+        {"topic": "test_topic", "timestamp": 12333, "key": 8, "value": {"GRADE": "F", "RANK": -1}},
+        {"topic": "test_topic", "timestamp": 12333, "key": 7, "value": {"GRADE": "F", "RANK": -2}},
+        {"topic": "test_topic", "timestamp": 12333, "key": 6, "value": {"GRADE": "F", "RANK": -3}}
+      ],
+      "responses": [
+        {"admin": {"@type": "currentStatus"}},
+        {"admin": {"@type": "currentStatus"}},
+        {"query": [
+          {"header":{"schema":"`GRADE` STRING KEY, `NUM_STUDENTS` BIGINT"}},
+          {"row":{"columns":["A",3]}},
+          {"row":{"columns":["F",4]}}
+
+        ]},
+        {"query": [
+          {"header":{"schema":"`GRADE` STRING KEY, `NUM_STUDENTS` BIGINT"}},
+          {"row":{"columns":["F",4]}}
+
+        ]},
+        {"query": [
+          {"header":{"schema":"`GRADE` STRING KEY"}},
+          {"row":{"columns":["F"]}}
+
+        ]}
+      ]
```

Comment:
Currently, there are no additional tests for Windowed materialization. Unsure if we need them right now

Reply:
You could always tested a windowed case like 
```
"CREATE TABLE AGGREGATE AS SELECT ID, COUNT(1) AS COUNT FROM INPUT WINDOW TUMBLING(SIZE 1 SECOND) GROUP BY ID HAVING COUNT(1) > 2;",
```
We don't have a lot of tests with the having clause for pull queries and I think it might trigger the `KsqlMaterialization` transform logic to apply the filtering.  Since they can apply for group bys I assume they work for windowed tables as well.  Normal where clauses presumably don't touch `KsqlMaterialization` when a group by is in place since the materialization happens after the filter has been applied, so that might not need additional testing.

But in general, you're right that a lot of windowed logic has already been tested fairly well in https://github.com/confluentinc/ksql/blob/master/ksqldb-functional-tests/src/test/resources/rest-query-validation-tests/pull-queries-against-materialized-aggregates.json since windows require a group by.

Reply:
These are good cases.  It would be good to do a windowed table + group by + having clause mentioning windowstart.  That's one last case I don't see.

Reply:
we don't allow `WINDOWSTART` that at the moment, so it's not a testable case. The specific error message is `Window bounds column WINDOWSTART can only be used in the SELECT clause of windowed aggregations and can not be passed to aggregate functions.`

Reply:
Related: https://github.com/confluentinc/ksql/issues/4397

Reply:
I'm still wondering, if @AlanConfluent 's query is possible:

```
CREATE TABLE AGGREGATE AS SELECT ID, COUNT(1) AS COUNT FROM INPUT WINDOW TUMBLING(SIZE 1 SECOND) GROUP BY ID HAVING COUNT(1) > 2;
```

note it does not try to group by window-start/end, while we should still be able to pull query it with conditions on other columns?